docs/2023/puzzles/day04.md

```diff
@@ -2,10 +2,184 @@ import Solver from "../../../../../website/src/components/Solver.js"
 
 # Day 4: Scratchcards
 
+by [@shardulc](https://github.com/shardulc)
+
 ## Puzzle description
 
 https://adventofcode.com/2023/day/4
 
+## Solution summary
+
+First, we note that both parts rely on counting how many winning numbers there
+are on a card, so we write the helper function `countWinning`. Then, part 1
+converts each card's winning numbers count to a number of points and adds them
+up. Part 2 cannot be expressed as a map over the winning counts because the
+value for a card possibly depends on the cards before it. Thus, we write it as a
+fold, where the accumulator tracks: (1) the number of cards won up to the
+current card; and (2) the number of copies of cards to be processed, i.e. their
+*multiplicities*, for as many of the following cards as needed.
+
+### Count number of winning numbers
+
+```scala
+def countWinning(card: String): Int =
+```
+
+Most of this function is string manipulation:
+
+```scala
+  val numbers = card
+    .substring(card.indexOf(":") + 1)   // discard "Card X:"
+    .split(" ")
+    .filterNot(_.isEmpty())
+  val (winningNumberStrs, givenNumberStrs) = numbers.span(_ != "|")
+  val winningNumbers = winningNumberStrs.map(_.toInt).toSet
+  // drop the initial "|"
+  val givenNumbers = givenNumberStrs.drop(1).map(_.toInt).toSet
+```
+
+Although this is not specified in the puzzle description, it seems like a number
+can appear at most once in the list of winning numbers as well as in the given
+numbers, so it is valid to perform `toSet` and the final counting step, which is
+what we return:
+
+```scala
+  winningNumbers.intersect(givenNumbers).size
+end countWinning
+```
+
+Lastly, the standard library conveniently gives us an iterator over lines.
+
+```scala
+def winningCounts(input: String): Iterator[Int] =
+  input.linesIterator.map(countWinning)
+```
+
+### Part 1
+
+```scala
+def part1(input: String): String =
+  winningCounts(input)
```

Comment:
I think its confusing because without explanation this looks like the whole code